### PR TITLE
CRL improvements and update callback

### PR DESCRIPTION
**Pull request by ColtonWilley**
# Description

CRL improvements, add parsing for CRL number, do not allow CRL duplicates, add callback for when CRL entry is updated.

# Checklist

 - [x] added tests
 - [x] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [x] Updated manual and documentation

### CODE REVIEW

File: src/ssl_certman.c [hunk 1]
```diff
@@ -1878,6 +1878,41 @@ int wolfSSL_CertManagerSetCRL_ErrorCb(WOLFSSL_CERT_MANAGER* cm, crlErrorCb cb,
...
 }
 
+#ifdef HAVE_CRL_UPDATE_CB
+int wolfSSL_CertManagerGetCRLInfo(WOLFSSL_CERT_MANAGER* cm, CrlInfo* info,
+    const byte* buff, long sz, int type)
+{
+    return GetCRLInfo(cm->crl, info, buff, sz, type);
+}
+
+/* Set the callback to be called when a CRL entry has
```
Comment (dgarske) [inline src/ssl_certman.c:1888]: Please make sure doxy is add in `doc/dox_comments/header_files/ssl.h` for new public API's. That's where the documentation is generated from.

File: src/crl.c [hunk 2]
```diff
@@ -560,12 +559,43 @@ int CheckCertCRL(WOLFSSL_CRL* crl, DecodedCert* cert)
             NULL, cert->extCrlInfo, cert->extCrlInfoSz, issuerName);
 }
 
+#ifdef HAVE_CRL_UPDATE_CB
+static void SetCrlInfo(CRL_Entry* entry, CrlInfo *info) {
```
Comment (dgarske) [inline src/crl.c:563]: Function start brace on new line please.

File: wolfssl/ssl.h [hunk 3]
```diff
@@ -3324,6 +3324,22 @@ typedef void (*CbOCSPRespFree)(void*,unsigned char*);
...
+typedef struct CrlInfo {
+    byte *issuerHash;
+    word32 issuerHashLen;
+    byte *lastDate;
+    word32 lastDateMaxLen;
+    byte lastDateFormat;
+    byte *nextDate;
+    word32 nextDateMaxLen;
+    byte nextDateFormat;
+    sword32 crlNumber;
```
Comment (dgarske) [inline wolfssl/ssl.h:3337]: Is/Should this really be signed 32-bit?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/ssl.h (hunk 3 above)
Comment (ColtonWilley) [reply wolfssl/ssl.h:3337]: So I am simply mapping to the existing types in the CRL_Entry struct. Even though it probably should be an unsigned int it is a signed int, so thats what I also used here.

Comment (dgarske) [reply wolfssl/ssl.h:3337]: Aren't those `int`?

Comment (ColtonWilley) [reply wolfssl/ssl.h:3337]: Yes it is an `int` in `CRL_Entry` but I was trying to use wolfssl types for the public facing `CrlInfo` struct.

File: wolfcrypt/src/asn.c [hunk 4]
```diff
@@ -38280,6 +38280,41 @@ static int ParseCRL_AuthKeyIdExt(const byte* input, int sz, DecodedCRL* dcrl)
...
+/* CRL NUM */          {0, ASN_INTEGER, 0, 0, 0 },
+};
+enum {
+    CRLNUMASN_IDX_CRL_NUM = 0
+};
+
+/* Number of items in ASN.1 template for CrlNumber. */
+#define crlNumASN_Length (sizeof(crlNumASN) / sizeof(ASNItem))
+
+static int ParseCRL_CrlNumExt(const byte* input, int sz, DecodedCRL* dcrl)
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:38295]: Use GetShortInt() instead?

**Review by SparkiDev — CHANGES_REQUESTED** [review]
Happy with the update code.

File: wolfcrypt/src/asn.c (hunk 4 above)
Comment (ColtonWilley) [reply wolfcrypt/src/asn.c:38295]: Yes that would be cleaner. I did not realize GetShortInt internals had a leg for WOLFSSL_ASN_TEMPLATE. I will rewrite to use GetShortInt instead.

File: src/crl.c [hunk 5]
```diff
@@ -560,12 +559,45 @@ int CheckCertCRL(WOLFSSL_CRL* crl, DecodedCert* cert)
             NULL, cert->extCrlInfo, cert->extCrlInfoSz, issuerName);
 }
 
+#ifdef HAVE_CRL_UPDATE_CB
+static void SetCrlInfo(CRL_Entry* entry, CrlInfo *info)
```
[the remark this answers is not in the document]
Comment (ColtonWilley) [reply src/crl.c:563]: Well the input structures are different right, CRL_entry vs DecodedCRL. I could make it void, pass a size and cast accordingly but I dont know if we have really gained much there vs the current implementation. If you strongly prefer it that way though let me know and I can make the changes.

[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/crl.c:563]: Both functions take a CRL_Entry and CrlInfo. Doing the same thing.

CRL_DIGEST_SIZE is the issuer hash size.
SIGNER_DIGEST_SIZE is the subject/issuer hash size.

Define CERT_NAME_HASH_SIZE and have the other two use it.

[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/crl.c:563]: Different type in first parameter. :-(
OK as is.

File: wolfcrypt/src/asn.c [hunk 6]
```diff
@@ -38472,7 +38471,16 @@ static int ParseCRL_Extensions(DecodedCRL* dcrl, const byte* buf, word32 idx,
                 }
             #endif
             }
-            /* TODO: Parse CRL Number extension */
+            else if (oid == CRL_NUMBER_OID) {
+                /* Parse CRL Number extension.
+                 * idx is at start of OCTET_STRING data. */
+                if (GetShortInt(buf,
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:38477]:
```
   Given the requirements above, CRL numbers can be expected to contain
   long integers.  CRL verifiers MUST be able to handle CRLNumber values
   up to 20 octets.  Conforming CRL issuers MUST NOT use CRLNumber
   values longer than 20 octets.
```
Need to use GetInt()

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Review by dgarske — CHANGES_REQUESTED** [review]
Please resolve merge conflicts.

File: src/crl.c [hunk 7]
```diff
@@ -686,6 +753,86 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
+            FreeDer(&der);
+            return -1;
+        }
+    #else
+        ret = NOT_COMPILED_IN;
+    #endif
+    }
+
+#ifdef WOLFSSL_SMALL_STACK
+    dcrl = (DecodedCRL*)XMALLOC(sizeof(DecodedCRL), NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (dgarske) [inline src/crl.c:794]: NIT: This line is over 80 characters.

```
overlong lines added:
src/crl.c:794     dcrl = (DecodedCRL*)XMALLOC(sizeof(DecodedCRL), NULL, DYNAMIC_TYPE_TMP_BUFFER);
```

File: wolfssl/ssl.h [hunk 8]
```diff
@@ -3491,6 +3491,22 @@ typedef void (*CbOCSPRespFree)(void*,unsigned char*);
...
+} CrlInfo;
+
+typedef void (*CbUpdateCRL)(CrlInfo* old, CrlInfo* new);
```
Comment (dgarske) [inline wolfssl/ssl.h:3507]: You cannot use `new` as a variable name with g++ compiler... 

```
configure:41488: g++-13 -c -DHAVE_CRL_UPDATE_CB   -Werror -g -ggdb -O0 -Wno-pragmas -Wall -Wextra -Wunknown-pragmas --param=ssp-buffer-size=1 -Waddress -Warray-bounds -Wchar-subscripts -Wcomment -Wfloat-equal -Wformat-security -Wformat=2 -Wmaybe-uninitialized -Wmissing-field-initializers -Wmissing-noreturn -Wnormalized=id -Wpointer-arith -Wshadow -Wsign-compare -Wstrict-overflow=1 -Wswitch-enum -Wundef -Wunused -Wunused-result -Wunused-variable -Wwrite-strings -fwrapv -DHAVE_CONFIG_H -I. -I. -DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK conftest.c >&5
In file included from ./wolfssl/openssl/ssl.h:37,
                 from ./wolfssl/openssl/aes.h:44,
                 from conftest.c:70:
./wolfssl/ssl.h:3507:52: error: expected ',' or '...' before 'new'
 3507 | typedef void (*CbUpdateCRL)(CrlInfo* old, CrlInfo* new);
      |                                                    ^~~
configure:41488: $? = 1
configure: failed program was:
```

File: src/crl.c [hunk 9]
```diff
@@ -560,12 +559,45 @@ int CheckCertCRL(WOLFSSL_CRL* crl, DecodedCert* cert)
...
 /* Add Decoded CRL, 0 on success */
 static int AddCRL(WOLFSSL_CRL* crl, DecodedCRL* dcrl, const byte* buff,
                   int verified)
 {
     CRL_Entry* crle = NULL;
+    CRL_Entry* curr = NULL;
+    CRL_Entry* prev = NULL;
+#ifdef HAVE_CRL_UPDATE_CB
+    CrlInfo old;
+    CrlInfo new;
```
Comment (dgarske) [inline src/crl.c:599]: Same here with g++:

```
[quantum-safe-wolfssl-all-g++-latest-debug] [6 of 36] [55be5035a0-dirty]
    autogen.sh 55be5035a0-dirty...   real 0m10.586s  user 0m8.680s  sys 0m1.088s
    configure...   real 0m25.038s  user 0m13.907s  sys 0m13.286s
    build...src/crl.c: In function ‘int AddCRL(WOLFSSL_CRL*, DecodedCRL*, const byte*, int)’:
src/crl.c:599:13: error: expected unqualified-id before ‘new’
  599 |     CrlInfo new;
```

File: tests/api.c [hunk 10]
```diff
@@ -89760,46 +89760,130 @@ static int test_wolfSSL_CTX_LoadCRL(void)
...
-    } test_params[] = {
-        { "./certs/server-cert.pem", "./certs/server-key.pem" },
-        { "./certs/server-revoked-cert.pem", "./certs/server-revoked-key.pem" }
-    };
-    size_t i;
+#if defined(HAVE_CRL) && !defined(NO_RSA) && !defined(NO_FILESYSTEM) && \
+    defined(HAVE_CRL_UPDATE_CB)
+int crlUpdateTestStatus = 0;
+WOLFSSL_CERT_MANAGER* updateCrlTestCm = NULL;
+static void updateCrlCb(CrlInfo* old, CrlInfo* new)
```
Comment (dgarske) [inline tests/api.c:89767]: Same here: 

```
[quantum-safe-wolfssl-all-g++-latest-debug] [6 of 36] [55be5035a0-dirty]
    autogen.sh 55be5035a0-dirty...   real 0m10.443s  user 0m8.567s  sys 0m1.048s
    configure...   real 0m25.272s  user 0m13.878s  sys 0m13.586s
    build...tests/api.c:89767:48: error: expected ‘,’ or ‘...’ before ‘new’
89767 | static void updateCrlCb(CrlInfo* old, CrlInfo* new)
      |                                                ^~~
tests/api.c: In function ‘void updateCrlCb(CrlInfo*, CrlInfo*)’:
tests/api.c:89782:28: error: expected type-specifier before ‘==’ token
89782 |     if (old == NULL || new == NULL) {
      |                            ^~
In file included from tests/api.c:32:
```

File: wolfssl/ssl.h [hunk 11]
```diff
@@ -3953,6 +3969,12 @@ WOLFSSL_API void wolfSSL_CTX_SetPerformTlsRecordProcessingCb(WOLFSSL_CTX* ctx,
     WOLFSSL_API int wolfSSL_CertManagerSetCRL_IOCb(WOLFSSL_CERT_MANAGER* cm,
         CbCrlIO cb);
 #endif
+#ifdef HAVE_CRL_UPDATE_CB
+    WOLFSSL_API int wolfSSL_CertManagerGetCRLInfo(WOLFSSL_CERT_MANAGER* cm, CrlInfo* info,
```
Comment (dgarske) [inline wolfssl/ssl.h:3973]: Optional fix. Note lots of other occurrences in this file, please don't worry about those.
```
overlong lines added:
wolfssl/ssl.h:3973     WOLFSSL_API int wolfSSL_CertManagerGetCRLInfo(WOLFSSL_CERT_MANAGER* cm, CrlInfo* info,
```

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**